Added initial onnx export function

nbertagnolli opened the pull request:
This PR is an initial attempt to address #78.  It creates a function called `export_onnx` which currently handles models with a sklearn based head. It will convert the body and the head to onnx format and merge them with a custom mean_pooling onnx layer in between.  I wanted to open this to get the conversation started I'd love feedback and hope to make this something generally useful : ).

Some notes:
1. I couldn't find an implementation of the attention based mean pooling anywhere so I made one. If there is already an implementation that would probably be better than what I put together.
2. I'm not sure how you want to handle the onnx dependencies now @lewtun Should we put them as a required pacakge to install? Should we do the imports in the export function and have the user custom install them if they need?
3. I owe you some tests 😅

blakechi commented:
Hi @nbertagnolli,

Thanks for creating the first attempt! For your first note, I think we can export the mean pooling layer together with the text encoder. I created a [notebook](https://drive.google.com/file/d/1KvN95i6vZlBW-eecq_waQiCwjO3vPvFR/view?usp=share_link) on my laptop to demo it. And I think by doing so, we can bypass the `transformers.convert_graph_to_onnx` issue you mentioned in your code.

Another thing is since we both need `model` and `pooler` in the notebook, how about we extract those two before [deciding `pytorch` or `sklearn` head](https://github.com/huggingface/setfit/pull/156/files#diff-32f86d0e3df2a295f2a7c5f6affc1626407031fc99e05c1af4d9423cf74a64b7R88)?

Please let me know what you think :)

blakechi commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,347 @@
+    model = SetFitModel.from_pretrained(model_name)
+
+    # Check to see if the model uses a sklearn head or a torch dense layer.
+    if issubclass(type(model.model_head), models.Dense):
```
How about before this if-else statement, we extract `model` and `pooler` as mentioned in this [notebook](https://drive.google.com/file/d/1KvN95i6vZlBW-eecq_waQiCwjO3vPvFR/view?usp=share_link)? So you can continue on connecting those two with the `sklearn` head and I can do so on the `pytorch` head @nbertagnolli

nbertagnolli replied:
This is awesome! Yeah pulling the pooler out and attaching them like this is way better.   I'll update the PR. Thanks! 
🙏 @blakechi 🙏

blakechi replied:
Glad you like it!
ok, I will add my park as well after your update

nbertagnolli replied:
I added in a bunch of your suggestions : ). Thanks for all the great input! Let me know if there is anything else you see that could be better.

blakechi replied:
Added my part at here: nbertagnolli/setfit#1

kgourgou commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,177 @@
+    """
+
+    # Load the model and get all of the parts.
+    model = SetFitModel.from_pretrained(model_name)
```
Looks really nice!

Recently I've recently been using more 

```
SetFitModel(body=SomeSentenceTransformer, head=SomeHead)
``` 

instead of "from_pretrained()". The advantage is that if I implement a model head in PyTorch, then I can easily spin a SetFit model with my custom head. I wonder if something like that makes sense here?

nbertagnolli replied:
That makes sense! I'll update this afternoon : )

kgourgou replied:
Thank you, it looks good!

kgourgou commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,212 @@
+    max_length = transformer.max_seq_length
+    model_body = transformer.auto_model
+    model_pooler = model_body_sentence_transformer._modules["1"]
+    model_head = model.model_head
```
Just highlighting this line; with the new commits here, there is no `model` variable anymore.

blakechi commented:
Hi @nbertagnolli , it seems like we have implemented the basic functionalities for exporting both kinds of heads. Would you like to refactor the `export_onnx` function such that some common snippets can be pulled out of the `if-else` statement to prevent duplication? If not, I would be more than happy to do it. 😄 So no pressure.

I find some parts we can improve as below, if you have others, feel free to share it as well

- `onnx.save`([here](https://github.com/huggingface/setfit/pull/156/files#diff-32f86d0e3df2a295f2a7c5f6affc1626407031fc99e05c1af4d9423cf74a64b7R174), [here](https://github.com/huggingface/setfit/pull/156/files#diff-32f86d0e3df2a295f2a7c5f6affc1626407031fc99e05c1af4d9423cf74a64b7R227)) can be pulled out of `if-else`.
- I suggest we add the tokenizer's metadata in the exported ONNX file for both heads such that users have a clue on which tokenizer they should get and how to use it (the [ONNX runtime API](https://github.com/huggingface/setfit/issues/78#issuecomment-1304593409) can also benefit from it)
- Wrap this [snippet](https://github.com/huggingface/setfit/pull/156/files#diff-32f86d0e3df2a295f2a7c5f6affc1626407031fc99e05c1af4d9423cf74a64b7R190-R205) into a function that accept necessary arguments

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    def __init__(
+        self,
+        model_body,
+        pooler,
```
I guess this can be a generic `Callable` right?

nbertagnolli replied:
I believe the type of the pooler is `sentence_transformers.models.Pooling.Pooling` which is a `nn.Module`. However,  we grab the "sentence_embedding" output from the pooling layer so a generic nn.Module wouldn't work.  What are your thoughts? Should we say the type of Pooler is Pooling or nn.Module and add some caveats in the docstring?

nbertagnolli replied:
Let me know if my solution addresses this comment : ).

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        if self.model_head is None:
+            return hidden
+
+        out = self.model_head(hidden)
```
Would this be more descriptive or does it only apply to the PyTorch model?
```suggestion
        logits = self.model_head(hidden)
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        hidden = {"token_embeddings": hidden[0], "attention_mask": attention_mask}
+
+        hidden = self.pooler(hidden)["sentence_embedding"]
+        if self.model_head is None:
```
Should we allow a user to export a model without a head? Just wondering whether to raise a `ValueError` here

nbertagnolli replied:
Sorry, this is confusing.  This OnnxSetFitModel is a wrapper that makes it easy to use torch.onnx.export to export the model_body with the pooling.  In the case of the torch based head we can use this to do the whole onnx export. However when we have an sklearn head we need to export the model_body with the pooler and then the sklearn head.  This was my attempt to make it easy to do just two onnx exports but then it means that the class name doesn't fully represent what is happening here.

lewtun replied:
Thanks, this makes sense!

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        hidden = self.model_body(input_ids, attention_mask, token_type_ids)
+        hidden = {"token_embeddings": hidden[0], "attention_mask": attention_mask}
+
+        hidden = self.pooler(hidden)["sentence_embedding"]
```
nit:
```suggestion
        embeddings = self.pooler(hidden)["sentence_embedding"]
```

Also is the `sentence_embedding` key necessary? I was thinking we could just require people to return a `torch.Tensor`

nbertagnolli replied:
Great suggestion! Yeah if we remove this key access then people can use other poolers more easily.

blakechi replied:
I think `sentence_embedding` is necessary since now we use `sentence-transformers`'s pooling layer.
Maybe instead of remove it, adding a if-else to check whether the pooler comes from `sentence-transformers`?

nbertagnolli replied:
I was thinking to make the type of Pooler `Optional[Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]]` Then to pass in the Pooler like we did before you could do:

```python
pooling_layer = model_body._modules["1"]
model_pooler = lambda x: pooling_layer(x)["sentence_embedding"]
OnnxSetFitModel(transformer, model_pooler, model_head).cpu()
```

This would let people pass in normal Callable's too. Thoughts?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        self.model_head = model_head
+
+    def forward(self, input_ids, attention_mask, token_type_ids):
+        hidden = self.model_body(input_ids, attention_mask, token_type_ids)
```
nit:
```suggestion
        hidden_states = self.model_body(input_ids, attention_mask, token_type_ids)
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+class ONNXSetFitModel(torch.nn.Module):
+    def __init__(
+        self,
+        model_body,
```
could we have some types here please :)

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    def __init__(
+        self,
+        model_body,
+        pooler,
```
What do you think about defining a function like this:

```python
def mean_pooling(token_embeddings, attention_mask) -> torch.Tensor:
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)
```

and using it as the default for the `pooler`? I took this from the model card of `sentence-transformers`: https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2#usage-huggingface-transformers

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        self.pooler = pooler
+        self.model_head = model_head
+
+    def forward(self, input_ids, attention_mask, token_type_ids):
```
Can we have some types here too please?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        return out
+
+
+def sklearn_head_to_onnx(model_head: LogisticRegression, opset: int):
```
nit:
```suggestion
def export_sklearn_head_to_onnx(model_head: LogisticRegression, opset: int):
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+
+def sklearn_head_to_onnx(model_head: LogisticRegression, opset: int):
+    """
+    Convert the sklearn head from a SetFitModel to ONNX format.
```
turbo nit:
```suggestion
    Convert the Scikit-Learn head from a SetFitModel to ONNX format.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    """
+    Convert the sklearn head from a SetFitModel to ONNX format.
+
+    :param model_head: The trained SetFit model_head.
```
Would you mind aligning the docstring format with those in the code base?

nbertagnolli replied:
Of course! Sorry 🙇

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        from sklearn.pipeline import Pipeline
+    except ImportError:
+        msg = (
+            "skl2onnx must be installed in order to convert a model with an sklearn head."
```
Can we provide the `pip` install command here to help the users?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        )
+        raise ImportError(msg)
+
+    # Check to see that the head has a coef_
```
should we also check if the head has been fitted or is that captured by this check as well?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    # If the datatype of the model is double we need to cast the outputs
+    # from the setfit model to doubles for compatibility inside of ONNX.
+    if type(dtype) == onnxconverter_common.data_types.DoubleTensorType:
+        # TODO:: TALK ABOUT FLOAT CONVERSION ISSUES.
```
Can you share what this refers to?

nbertagnolli replied:
Revisiting this I don't think it's an issue since we are going float to double. I was worried about doing a type conversion with floats and doubles.  If you used something like a random forest here and did a double to float conversion you could have really weird results since the function isn't continuous.  See this for more details. https://onnx.ai/sklearn-onnx/auto_tutorial/plot_ebegin_float_double.html

I'm going to remove this todo because since we are going from float to double we're increasing precision not decreasing it 😅.

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    ignore_ir_version: bool = True,
+):
+    """
+    Export a PyTorch backed setfit model to ONNX Intermediate Representation.
```
```suggestion
    Export a PyTorch backed SetFit model to ONNX Intermediate Representation.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+
+    Args:
+        model_body_sentence_transformer: The model_body from a SetFitModel.model_body. This should be a
+            SentenceTransformer.
```
I think we can indicate this in the typing right?

nbertagnolli replied:
yessir! Done : )

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    Args:
+        model_body_sentence_transformer: The model_body from a SetFitModel.model_body. This should be a
+            SentenceTransformer.
+        model_head: The model_head from SetFitModel.model_head. This can be either a dense
```
same comment as above:
```suggestion
        model_head: The SetFit model head. This can be either a dense
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    """
+    Export a PyTorch backed setfit model to ONNX Intermediate Representation.
+
+    Args:
```
Could we add some types to these args please?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+            might be lower. ONNX will use the highest version supported by both the
+            sklearn head and the model body. If versions can't be rectified an error
+            will be thrown.
+        output: The path where will be stored the generated ONNX model. If the head is an
```
nit:
```suggestion
        output_path: The path where will be stored the generated ONNX model. If the head is an
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+    )
+    dummy_sample = "It's a test."
+    dummy_inputs = tokenizer(dummy_sample, **tokenizer_kwargs)
+    symbolic_names = {0: "batch_size", 1: "max_seq_len"}
```
Maybe it's a bit clearer to do something like the following?

```python
input_names = list(dummy_inputs.keys())
output_names = ["logits"]
# dynamic axes
dynamic_axes_input = {}
for input_name in input_names:
    dynamic_axes_input[input_name] = {0: "batch_size", 1: "sequence"}

dynamic_axes_output = {}
for output_name in output_names:
    dynamic_axes_output[output_name] = {0: "batch_size", 1: "sequence"}

dynamic_axes = {**dynamic_axes_input, **dynamic_axes_output}
```

nbertagnolli replied:
I pulled some of this out into a function let me know if it's clearer now : ).

lewtun reviewed (commented):
Thanks a lot for this advanced feature @nbertagnolli - very cool work 🔥 !!

I've left a few comments, but the general API is looking quite good. Would you mind sharing a few code snippets showing how this works in practice?

nbertagnolli commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+from sklearn.linear_model import LogisticRegression
+
+
+class ONNXSetFitModel(torch.nn.Module):
```
Thanks for the comments @lewtun ! I'll start working through them : ).

For the usage examples, would you like those in the README, in a notebook under notebooks, or both? : )

lewtun replied:
Let's start with a notebook for now as the README is getting rather long :)

I've opened an issue https://github.com/huggingface/setfit/issues/181 to have real documentation and we can add the ONNX details there once it's up

nbertagnolli commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,227 @@
+        sk_model = model_head
+
+    # Convert sklearn head into ONNX format
+    initial_type = [("model_head", dtype)]
```
totes!

nbertagnolli commented:
Thanks for all of the great feedback everyone! I really appreciate all of the insightful comments : ). I think 🤞 I've addressed all of the comments. Please let me know if I missed anything and I'll take another crack at it.

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+from transformers.modeling_utils import PreTrainedModel
+
+
+def mean_pooling(token_embeddings: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
```
Very nice!

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+
+
+def mean_pooling(token_embeddings: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
+    """Perform attention aware mean pooling.
```
turbo nit:
```suggestion
    """Perform attention-aware mean pooling.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+    pooling across the sequence dimension to yield embeddings of size (batch, embedding_size).
+
+    From:
+    https://huggingface.co/sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2#usage-huggingface-transformers
```
Let's cite the source in case the model repo changes :)
```suggestion
    https://github.com/UKPLab/sentence-transformers/blob/0b5ef4be93d2b21de3a918a084b48aab6ba48595/sentence_transformers/model_card_templates.py#L134
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+class OnnxSetFitModel(torch.nn.Module):
+    """A wrapper around SetFit model body, pooler, and model head which makes ONNX exporting easier.
+
+    This wrapper allows us to create a nn.Module with different levels of connectivity. We can set
```
nit:
```suggestion
    This wrapper creates a nn.Module with different levels of connectivity. We can set
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+    """A wrapper around SetFit model body, pooler, and model head which makes ONNX exporting easier.
+
+    This wrapper allows us to create a nn.Module with different levels of connectivity. We can set
+    model_body and pooler and have a Module which maps inputs to embeddings or we can set all three
```
```suggestion
    `model_body` and `pooler` and have a Module which maps inputs to embeddings or we can set all three
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+
+    This wrapper allows us to create a nn.Module with different levels of connectivity. We can set
+    model_body and pooler and have a Module which maps inputs to embeddings or we can set all three
+    and have a model which maps inputs to final predictions. This is useful because torch.onnx.export
```
```suggestion
    and have a model which maps inputs to final predictions. This is useful because `torch.onnx.export`
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+    This wrapper allows us to create a nn.Module with different levels of connectivity. We can set
+    model_body and pooler and have a Module which maps inputs to embeddings or we can set all three
+    and have a model which maps inputs to final predictions. This is useful because torch.onnx.export
+    will work with a nn.Module.
```
```suggestion
    will work with a `nn.Module`.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+
+    Attributes:
+        model_body (`PreTrainedModel`): The pretrained model body of a setfit model.
+        pooler (`nn.Module` or `Callable[[torch.Tensor], torch.Tensor]`): The callable function that can
```
I think we can use `Union` here for these types
```suggestion
        pooler (`Union[nn.Module, Callable[[torch.Tensor], torch.Tensor]]`, *optional*, defaults to `None`): The callable function that can
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+        model_body (`PreTrainedModel`): The pretrained model body of a setfit model.
+        pooler (`nn.Module` or `Callable[[torch.Tensor], torch.Tensor]`): The callable function that can
+            map  tensors of shape (batch, sequence, embedding_dim) to shape (batch, embedding_dim).
+        model_head: (`torch.nn.Module` or `LogisticRegression`): The model head from the pretrained SetFit model.
```
Use `nn.Module` for consistency with above?
```suggestion
        model_head: (`Union[nn.Module, LogisticRegression]`, *optional*, defaults to `None`): The model head from the pretrained SetFit model.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+        pooler (`nn.Module` or `Callable[[torch.Tensor], torch.Tensor]`): The callable function that can
+            map  tensors of shape (batch, sequence, embedding_dim) to shape (batch, embedding_dim).
+        model_head: (`torch.nn.Module` or `LogisticRegression`): The model head from the pretrained SetFit model.
+            if it is None then the resulting OnnxSetFitModel forward will return embeddings instead of predictions.
```
```suggestion
            If `None`, then the resulting `OnnxSetFitModel.forward` method will return embeddings instead of predictions.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+
+        self.model_body = model_body
+        if pooler is None:
+            self.pooler = mean_pooling
```
Maybe we can use the logger here to indicate the no pooler was provided, so we're defaulting to mean pooling?

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+def export_onnx_setfit_model(setfit_model: OnnxSetFitModel, inputs, output_path, opset):
+    """Export the OnnxSetFitModel.
+
+    This exports the model created by the OnnxSetFitModel wrapper using torch.onnx.export.
```
```suggestion
    This exports the model created by the `OnnxSetFitModel` wrapper using `torch.onnx.export`.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+
+
+def export_onnx_setfit_model(setfit_model: OnnxSetFitModel, inputs, output_path, opset):
+    """Export the OnnxSetFitModel.
```
```suggestion
    """Export the `OnnxSetFitModel`.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+        inputs (`Dict[str, torch.Tensor]`): The inputs we would hypothetically pass to the model. These are
+            generated using a tokenizer.
+        output_path (`str`): The local path to save the onnx model to.
+        opset (`int`): The ONNX opset to use for the export.
```
What do you think about setting a good default value for `opset` since it might be annoying for users to have to look it up

nbertagnolli replied:
That's a great suggestion! I've had 12 work well so I'll set that to start : )

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+    Args:
+        model_head (`LogisticRegression`): The trained SetFit model_head.
+        opset (`int`): The ONNX opset to use for optimizing this model. The opset is not
+            gauranteed and will default to the maximum version possible for the sklearn
```
```suggestion
            guaranteed and will default to the maximum version possible for the sklearn
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+        [`onnx.onnx_ml_pb2.ModelProto`] The ONNX model generated from the sklearn head.
+
+    Raises:
+        ImportError: If skl2onnx is not installed an error will be raised asking
```
```suggestion
        ImportError: If `skl2onnx` is not installed an error will be raised asking
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+    except ImportError:
+        msg = """
+        skl2onnx must be installed in order to convert a model with an sklearn head.
+        Please install with pip install skl2onnx.
```
```suggestion
        Please install with `pip install skl2onnx`.
```

lewtun commented on src/setfit/onnx.py:
```diff
@@ -0,0 +1,281 @@
+        from sklearn.pipeline import Pipeline
+    except ImportError:
+        msg = """
+        skl2onnx must be installed in order to convert a model with an sklearn head.
```
```suggestion
        `skl2onnx` must be installed in order to convert a model with an sklearn head.
```

lewtun reviewed (commented):
Thanks a lot for iterating on this feature @nbertagnolli - it's looking very nice 🚀 !

I've left some nits, but apart from that I think the only missing piece is some unit tests. You could pick a small checkpoint from the Hub and verify that the forward pass of the model in PyTorch vs ONNX Runtime is the same (within some tolerance)

nbertagnolli commented:
Sorry about the delay was out a bit last week.  I think I've addressed that next round of comments : ). I also added a small notebook demonstrating the usage of `export_onnx` function and a basic test to ensure that the models outout matches the original model. Please let me know if there is anything else you notice we should improve :D.

nbertagnolli commented:
@lewtun I'm not sure how you'd like to manage dependencies in setup.py. I added them to the `TESTS_REQUIRE` dependencies list because they are needed for testing. I don't know if that makes the most sense or if there is something better we could do.

nbertagnolli commented:
I found the problem : ), I wasn't getting the maximum opset for the sklearn head! Yay tests!

lewtun commented:
Sorry for the delay in reviewing this @nbertagnolli - on it right now!

lewtun reviewed (approved):
Thanks a lot for iterating on this killer feature @nbertagnolli - it's looking great!

I've left some tiny nits and a suggestion to move the files into a dedicated `setfit.exporters` module so that we can easily support new exporters like OpenVINO.

After that we can finally merge this beast 🔥 !

nbertagnolli commented:
Thanks for the great comments as always! I made the new exporters module and put onnx in there : ). Let me know if you see anything else!

tomaarsen commented:
I'm missing a `setfit/exporters/__init__.py` to turn `setfit.exporters` into a module. Furthermore, `setfit/exporters/utils.py` is not included, despite an attempt to import from it.

- Tom Aarsen

lewtun commented:
Hey @nbertagnolli thanks for the last changes, would you mind running `make style && make quality` locally and pushing the formatted code? I think that should fix the CI :)

nbertagnolli commented:
Sorry about all the style commits that's what i get for being lazy and not running the style checker locally 😅 😅.

lewtun commented:
Wonderful, thank you so much for this 🔥 !!
